src/kdc/do_tgs_req.c

```diff
@@ -417,10 +420,15 @@ process_tgs_req(krb5_kdc_req *request, krb5_data *pkt,
         }
     }
 
-    if (is_referral)
-        ticket_reply.server = server->princ;
+    if (is_referral || (isflagset(request->kdc_options, KDC_OPT_CANONICALIZE) &&
+                        krb5_is_tgs_principal(server->princ)))
```

Comment:
I'm curious about this change.  It only has an effect if is_referral is false but server->princ is a TGS principal and is different from request->server.  is_referral is set by:

    is_referral = is_cross_tgs_principal(server->princ) &&
        !krb5_principal_compare(kdc_context, request->server, server->princ);

So we can infer that server->princ is a local TGS principal; that is, the client requested krbtgt/ALIAS@ALIAS (or krbtgt/ALIAS@REALM) and the KDB returned krbtgt/REALM@REALM.  I have two questions:

1. Is this a likely request, or just a theoretical point of compatibility with Windows?  This is the TGS path, so the client has already had a chance to canonicalize the realm.

2. Can we handle this more simply by just setting is_referral to true?  (That is, when we set is_referral, use krb5_is_tgs_principal() instead of is_cross_tgs_principal().)  It would be the KDB's responsibility not to canonicalize the TGS name if KRB5_KDB_FLAG_CANONICALIZE isn't set.

Reply:
Oh, but we now say that modules can always return in-realm aliases, so we'd be going back on that.  So I guess I just have the first question.

Reply:
I think it is likely, it happens for instance in the capture I attached to #1110, packets 80-84:
https://github.com/iboukris/S4U/blob/master/logs/xrealm_s4u2self.pcapng

And yeah, it is simpler when the kdb always return the canonical name, i'll look at the code again.

Reply:
I changed the condition a bit but I can't think of anything better, let me know if you prefer option 2. and we'll have to make an exception in samba for krbtgt alias lookup when the canon flag isn't set.

Reply:
> and we'll have to make an exception in samba for krbtgt alias lookup when the canon flag isn't set.

I'm not following; why does changing the condition to use is_local_tgs_principal() require a KDB change?

Reply:
That was unrelated to the is_local_tgs_principal() change, if we go with option 2. from the ones you gave above, that is always canonicalize krbtgt reply regardless of the the canon flag.

Reply:
Understood.  We'll stick with this method rather than making the KDB requirements more complicated.